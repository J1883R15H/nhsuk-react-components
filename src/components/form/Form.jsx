import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import FormContext from './FormContext';
import { Heading } from '../typography';

class Form extends Component {
  static propTypes = {
    className: PropTypes.string,
    onChange: PropTypes.func,
    onSubmit: PropTypes.func,
    children: PropTypes.node.isRequired,
    onErrorChange: PropTypes.func
  };

  static defaultProps = {
    className: '',
    onChange: () => {},
    onSubmit: () => {},
    onErrorChange: () => {}
  };

  constructor(props) {
    super(props);
    this.state = {
      formData: {},
      erroredComponents: {}
    };
  }

  registerComponent = (name, value) => {
    this.setState(() => {
      const { formData } = this.state;
      formData[name] = value;
      return { formData };
    });
  };

  updateFormState = (name, value) => {
    const { onChange } = this.props;
    this.setState(() => {
      const { formData } = this.state;
      formData[name] = value;
      if (onChange) onChange(formData);
      return { formData };
    });
  };

  onSubmit = e => {
    const { onSubmit } = this.props;
    const { formData } = this.state;
    if (onSubmit) onSubmit(e, formData);
  };

  passBackError = (name, error = true, errorDetails) => {
    const { erroredComponents } = this.state;
    const original = erroredComponents;
    if (error) {
      erroredComponents[name] = errorDetails;
    } else if (Object.keys(erroredComponents).includes(name)) {
      delete erroredComponents[name];
    }

    if (original !== erroredComponents) {
      this.setState({ erroredComponents }, () => {
        const { onErrorChange } = this.props;
        onErrorChange(erroredComponents);
      });
    }
  };

  render() {
    const {
      className,
      children,
      action,
      method,
      title,
      titleSize,
      ...rest
    } = this.props;
    const { erroredComponents } = this.state;
    const contextValue = {
      registerComponent: this.registerComponent,
      updateFormState: this.updateFormState,
      passBackError: this.passBackError
    };
    return (
      <form action={action} method={method} onSubmit={this.onSubmit}>
        <div
          className={classNames(
            'nhsuk-form-group',
            { 'nhsuk-form-group--error': erroredComponents.length > 0 },
            className
          )}
          {...rest}
        >
          {title ? <Heading size={titleSize}>{title}</Heading> : null}
          <FormContext.Provider value={contextValue}>
            {children}
          </FormContext.Provider>
        </div>
      </form>
    );
  }
}

export default Form;
