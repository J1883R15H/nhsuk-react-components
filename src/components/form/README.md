# Form

The FormWrapper renders only a basic wrapper for the form, but it contains functions that are automatically passed to other form components. This allows the Form object to act as a central state store for the form, and exposing the onSubmit and onChange functions centrally for the whole form. All NHSUK React Form components will automatically "call-back" and register with the central form object, updating it's state. They also automatically pass down changes to the central form object, so everything should work fairly intuitively.

## Technical Documentation

When the Form renders, it maps over it's children and checks for any NHSUK React Form components. If it finds them, it creates a clone of that component, with the additional `valueCallback` and `registerInitialValue` functions, as well as some helpers such as `elementMargin` which sets up spacing for the components. The NHSUK React Form components then call the `registerInitialValue` function with the form element name and the initial value of that element. On every change, they call the `valueCallback` function, again with the name and initial value.