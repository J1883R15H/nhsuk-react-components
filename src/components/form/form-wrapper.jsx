import React from 'react';
import PropTypes from 'prop-types';
import stylePropType from 'react-style-proptype';
import classNames from 'classnames';
import Heading from '../typography/heading';

class Form extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      formData: {},
      errorInChild: false
    };
    this.valueCallback = this.valueCallback.bind(this);
    this.registerInitialValue = this.registerInitialValue.bind(this);
    this.callbackErrorFromChild = this.callbackErrorFromChild.bind(this);
  }

  componentWillMount() {
    const { children } = this.props;
    const { errorInChild } = this.state;
    React.Children.forEach(children, child => {
      const { props } = child;
      if (props) {
        if (props.error && !errorInChild) {
          this.setState({ errorInChild: true });
        }
      }
    });
  }

  registerInitialValue(key, value) {
    this.setState(previousState => {
      const previousFormData = previousState.formData;
      return { formData: { ...previousFormData, [key]: value } };
    });
  }

  valueCallback(key, value) {
    const { onChange } = this.props;
    this.setState(previousState => {
      const previousFormData = previousState.formData;
      onChange({ formData: { ...previousFormData, [key]: value } });
      return { formData: { ...previousFormData, [key]: value } };
    });
  }

  callbackErrorFromChild(errorInChild) {
    this.setState({ errorInChild });
  }

  render() {
    const {
      submitAction,
      submitMethod,
      children,
      style,
      className,
      title,
      titleSize,
      elementMargin,
      onSubmit
    } = this.props;
    const { formData } = this.state;
    const { errorInChild } = this.state;
    const modifiedChildren = React.Children.map(children, child => {
      const { props } = child;
      if (props) {
        const formElements = [
          'Checkboxes',
          'DateInput',
          'Input',
          'Fieldset',
          'InputBlock',
          'Radios',
          'Select',
          'Textarea'
        ];
        if (formElements.includes(child.type.name)) {
          return React.cloneElement(child, {
            valueCallback: this.valueCallback,
            style: { ...child.props.style, marginBottom: elementMargin },
            elementMargin,
            registerInitialValue: this.registerInitialValue,
            callbackErrorFromChild: this.callbackErrorFromChild
          });
        }
        return child;
      }
      return child;
    });
    return (
      <form
        action={submitAction}
        method={submitMethod}
        onSubmit={e => onSubmit(e, formData)}
      >
        {title ? <Heading size={titleSize}>{title}</Heading> : null}
        <div
          className={classNames(
            'nhsuk-form-group',
            { 'nhsuk-form-group--error': errorInChild },
            className
          )}
          style={style}
        >
          {modifiedChildren}
        </div>
      </form>
    );
  }
}

Form.propTypes = {
  submitAction: PropTypes.string,
  submitMethod: PropTypes.string,
  children: PropTypes.node.isRequired,
  onSubmit: PropTypes.func,
  className: PropTypes.string,
  style: stylePropType,
  elementMargin: PropTypes.number,
  onChange: PropTypes.func,
  title: PropTypes.string,
  titleSize: PropTypes.string
};

Form.defaultProps = {
  submitAction: '',
  submitMethod: 'GET',
  className: '',
  style: {},
  elementMargin: 0,
  onSubmit: () => {},
  onChange: () => {},
  title: '',
  titleSize: 'm'
};

export default Form;
