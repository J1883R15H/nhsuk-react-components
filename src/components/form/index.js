import Form from './form-wrapper';
import Checkboxes from './components/checkboxes';
import DateInput from './components/date-input';
import Input from './components/input';
import Radios from './components/radios';
import Select from './components/select';
import Textarea from './components/textarea';
import InputBlock from './components/input-block';

export {
  Form,
  Checkboxes,
  DateInput,
  Input,
  Radios,
  Select,
  Textarea,
  InputBlock
};
