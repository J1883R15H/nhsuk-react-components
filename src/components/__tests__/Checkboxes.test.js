import React from 'react';
import { mount, shallow } from 'enzyme';
import sinon from 'sinon';
import { Checkboxes } from '../form';

describe('Checkboxes', () => {
  it('box matches snapshot', () => {
    const shallowMount = shallow(
      <Checkboxes.Box name="checkboxes" _id="id" value="option">
        Option
      </Checkboxes.Box>
    );
    expect(shallowMount).toMatchSnapshot();
    shallowMount.unmount();
  });
  it('box renders hint', () => {
    const fullMount = mount(
      <Checkboxes.Box name="checkboxes" _id="id" value="option" hint="hint">
        Option
      </Checkboxes.Box>
    );
    expect(
      fullMount.containsMatchingElement(
        <span className="nhsuk-hint nhsuk-checkboxes__hint">hint</span>
      )
    ).toBeTruthy();
    fullMount.unmount();
  });
  it('box disables when prop supplied', () => {
    const shallowMount = shallow(
      <Checkboxes.Box name="checkboxes" disabled _id="id" value="option">
        Option
      </Checkboxes.Box>
    );
    expect(shallowMount.find('input').props().disabled).toBe(true);
    shallowMount.unmount();
  });
  it('box handles onClick', () => {
    const onClick = sinon.spy();
    const fullMount = mount(
      <Checkboxes.Box
        name="checkboxes"
        _id="id"
        _onClick={onClick}
        value="option"
        hint="hint"
      >
        Option
      </Checkboxes.Box>
    );
    const fullMountForAnonFunc = mount(
      <Checkboxes.Box name="checkboxes" _id="id" value="option" hint="hint">
        Option
      </Checkboxes.Box>
    );
    fullMount.find('input').simulate('click');
    fullMountForAnonFunc.find('input').simulate('click');
    sinon.assert.called(onClick);
    fullMount.unmount();
  });
  it('wrapper matches snapshot', () => {
    const fullMount = mount(
      <Checkboxes name="name">
        <Checkboxes.Box value="box">Box</Checkboxes.Box>
      </Checkboxes>
    );
    expect(fullMount).toMatchSnapshot();
    fullMount.unmount();
  });
  it('wrapper catches details from boxes', () => {
    const props = {
      registerInitialValue: sinon.spy(),
      valueCallback: sinon.spy(),
      name: 'name'
    };
    const fullMount = mount(
      <Checkboxes {...props}>
        <Checkboxes.Box value="box">Box</Checkboxes.Box>
        <Checkboxes.Box value="box2">Box2</Checkboxes.Box>
      </Checkboxes>
    );
    const fullMountForAnonFunc = mount(
      <Checkboxes name="name">
        <Checkboxes.Box value="box">Box</Checkboxes.Box>
        <Checkboxes.Box value="box2">Box2</Checkboxes.Box>
      </Checkboxes>
    );
    expect(fullMount.state()).toEqual({ currentValue: [] });
    sinon.assert.calledOnce(props.registerInitialValue);
    sinon.assert.notCalled(props.valueCallback);
    fullMount.find('#name-1').simulate('click');
    fullMountForAnonFunc.find('#name-1').simulate('click');
    sinon.assert.calledOnce(props.valueCallback);
    expect(fullMount.state()).toEqual({ currentValue: ['box'] });
    fullMount.find('#name-2').simulate('click');
    sinon.assert.calledTwice(props.valueCallback);
    expect(fullMount.state()).toEqual({ currentValue: ['box', 'box2'] });
    fullMount.find('#name-1').simulate('click');
    sinon.assert.calledThrice(props.valueCallback);
    expect(fullMount.state()).toEqual({ currentValue: ['box2'] });
    fullMount.find('#name-2').simulate('click');
    sinon.assert.callCount(props.valueCallback, 4);
    expect(fullMount.state()).toEqual({ currentValue: [] });
    sinon.assert.calledOnce(props.registerInitialValue);
    fullMount.unmount();
  });
  it('returns other children normally', () => {
    const fullMount = mount(
      <Checkboxes name="name">
        <p>Hello</p>
      </Checkboxes>
    );
    expect(fullMount).toMatchSnapshot();
    fullMount.unmount();
  });
  it('renders with hint and error', () => {
    const mountwithError = mount(
      <Checkboxes name="name" error="error">
        <Checkboxes.Box value="box">Box</Checkboxes.Box>
        <Checkboxes.Box value="box2">Box2</Checkboxes.Box>
      </Checkboxes>
    );
    const mountwithHint = mount(
      <Checkboxes name="name" hint="hint">
        <Checkboxes.Box value="box">Box</Checkboxes.Box>
        <Checkboxes.Box value="box2">Box2</Checkboxes.Box>
      </Checkboxes>
    );
    expect(
      mountwithHint.containsMatchingElement(
        <span className="nhsuk-hint">hint</span>
      )
    ).toBeTruthy();

    expect(
      mountwithError.containsMatchingElement(
        <span className="nhsuk-error-message">
          <span className="nhsuk-u-visually-hidden">Error: </span>error
        </span>
      )
    ).toBeTruthy();
  });
});
