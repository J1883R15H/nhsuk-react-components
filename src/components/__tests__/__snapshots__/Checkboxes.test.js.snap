// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkboxes box matches snapshot 1`] = `
<div
  className="nhsuk-checkboxes__item"
  style={Object {}}
>
  <input
    aria-describedby="id__label"
    className="nhsuk-checkboxes__input"
    disabled={false}
    id="id"
    name="checkboxes"
    onClick={[Function]}
    type="checkbox"
    value="option"
  />
  <Label
    className="nhsuk-checkboxes__label"
    htmlFor="id"
    style={Object {}}
    type=""
  >
    Option
  </Label>
</div>
`;

exports[`Checkboxes returns other children normally 1`] = `
<Checkboxes
  className=""
  error=""
  hint=""
  label=""
  labelHtmlFor=""
  name="name"
  registerInitialValue={[Function]}
  style={Object {}}
  valueCallback={[Function]}
>
  <div
    className="nhsuk-checkboxes"
    style={Object {}}
  >
    <p
      key=".0"
    >
      Hello
    </p>
  </div>
</Checkboxes>
`;

exports[`Checkboxes wrapper matches snapshot 1`] = `
<Checkboxes
  className=""
  error=""
  hint=""
  label=""
  labelHtmlFor=""
  name="name"
  registerInitialValue={[Function]}
  style={Object {}}
  valueCallback={[Function]}
>
  <div
    className="nhsuk-checkboxes"
    style={Object {}}
  >
    <Box
      _id="name-1"
      _onClick={[Function]}
      className=""
      disabled={false}
      hint=""
      key=".0"
      name="name"
      style={Object {}}
      value="box"
    >
      <div
        className="nhsuk-checkboxes__item"
        style={Object {}}
      >
        <input
          aria-describedby="name-1__label"
          className="nhsuk-checkboxes__input"
          disabled={false}
          id="name-1"
          name="name"
          onClick={[Function]}
          type="checkbox"
          value="box"
        />
        <Label
          className="nhsuk-checkboxes__label"
          htmlFor="name-1"
          style={Object {}}
          type=""
        >
          <label
            className="nhsuk-label nhsuk-checkboxes__label"
            htmlFor="name-1"
            id="name-1__label"
            style={Object {}}
          >
            Box
          </label>
        </Label>
      </div>
    </Box>
  </div>
</Checkboxes>
`;
